File: test/Benzene.py
# compare Benzene result with pyscf

from time import time_ns
import numpy as np
from pyscf import gto
from pyscf import scf

a0 = 0.529_177_210_903
hartree = 27.211386245988

def Benzene(dCC:float, dCH:float):
    sqrt3_half = np.sqrt(3)/2
    RH = dCC + dCH
    return gto.M(
        atom = [
            ['C', np.array([  1.,          0., 0.]) * dCC],
            ['C', np.array([ 0.5,  sqrt3_half, 0.]) * dCC],
            ['C', np.array([-0.5,  sqrt3_half, 0.]) * dCC],
            ['C', np.array([ -1.,          0., 0.]) * dCC],
            ['C', np.array([-0.5, -sqrt3_half, 0.]) * dCC],
            ['C', np.array([ 0.5, -sqrt3_half, 0.]) * dCC],
            ['H', np.array([  1.,          0., 0.]) * RH],
            ['H', np.array([ 0.5,  sqrt3_half, 0.]) * RH],
            ['H', np.array([-0.5,  sqrt3_half, 0.]) * RH],
            ['H', np.array([ -1.,          0., 0.]) * RH],
            ['H', np.array([-0.5, -sqrt3_half, 0.]) * RH],
            ['H', np.array([ 0.5, -sqrt3_half, 0.]) * RH],
        ],
        basis = "sto-3g",
        verbose = 0
    )

def Benzene_pes(dCC_range, dCH_range):
    NC = len(dCC_range)
    NH = len(dCH_range)
    Emat = np.zeros((NC, NH))
    for i in range(NC):
        for j in range(NH):
            ben = Benzene(dCC_range[i], dCH_range[j])
            hf = scf.RHF(ben)
            hf.chkfile = None
            Emat[i,j] = hf.kernel()
    pos = np.argmin(Emat)
    i, j = pos // NH, pos % NH
    print(f"minimal energy is {Emat[i,j]*hartree}eV, with dCC = {dCC_range[i]:.2f}Å, dCH = {dCH_range[j]:.2f}Å")

if __name__ == "__main__":
    start = time_ns()
    Benzene_pes(np.linspace(1.3, 1.5, 11), np.linspace(1, 1.2, 11))
    stop = time_ns()
    print("time used:", (stop-start)/1e9)

File: test/Benzene.jl
include("../src/HartreeFock.jl")

using .HartreeFock
using BenchmarkTools

# define some unit vector
const n = (
    vec3(  1.,    0., 0.),
    vec3( 1/2,  √3/2, 0.),
    vec3(-1/2,  √3/2, 0.),
    vec3( -1.,    0., 0.),
    vec3(-1/2, -√3/2, 0.),
    vec3( 1/2, -√3/2, 0.)
)

# dCC: bond length between adjacent `C` Atom
# dCH: bond length between `C` and `H`
function Benzene(dCC::Float64, dCH::Float64)::Molecule
    Molecule("C6H6", vcat(
        collect(ntuple(i->Atom("C", dCC * n[i]), 6)),
        collect(ntuple(i->Atom("H", (dCC+dCH) * n[i]), 6))
    ))
end

function Benzene_pes(dCC_range, dCH_range)
    NCC = length(dCC_range)
    NCH = length(dCH_range)
    Emat = zeros(Float64, NCC, NCH)
    Threads.@threads for i in 1:NCC
    # for i in 1:NCC
        for j in 1:NCH
            benzene = Benzene(dCC_range[i] / a0, dCH_range[j] / a0)
            total_energy = hartree_fock(benzene)
            Emat[i, j] = total_energy
        end
    end
    i, j = argmin(Emat).I
    println("minimal energy is $(Emat[i,j]), with dCC = $(dCC_range[i])Å, dCH = $(dCH_range[j])Å")
end

#display(@benchmark Benzene_pes(1.3:0.02:1.5, 1:0.02:1.2))
Benzene_pes(1.3:0.02:1.5, 1:0.02:1.2)